modules/network/src/main/java/org/apache/ignite/internal/network/direct/stream/DirectByteBufferStreamImplV1.java

```diff
@@ -496,6 +517,89 @@ public void writeBitSet(BitSet val) {
         writeLongArray(val != null ? val.toLongArray() : null);
     }
 
+    /** {@inheritDoc} */
+    @Override
+    public void writeByteBuffer(ByteBuffer val) {
+        switch (byteBufferState) {
+            case 0:
+                byte flag = 0;
+
+                if (val != null) {
+                    flag |= BYTE_BYFFER_NOT_NULL_FLAG;
+
+                    if (val.order() == ByteOrder.BIG_ENDIAN) {
+                        flag |= BYTE_BYFFER_BIG_ENDIAN_FLAG;
+                    }
+
+                    if (val.position() != 0) {
+                        flag |= BYTE_BYFFER_HAS_POSITION_FLAG;
+                    }
+
+                    if (val.limit() != val.capacity()) {
+                        flag |= BYTE_BYFFER_HAS_LIMIT_FLAG;
+                    }
+                }
+
+                writeByte(flag);
+
+                if (!lastFinished || val == null) {
+                    return;
+                }
+
+                byteBufferState++;
+
+                //noinspection fallthrough
+            case 1:
+                if (val.position() != 0) {
+                    writeInt(val.position());
+                }
+
+                if (!lastFinished) {
+                    return;
+                }
+
+                byteBufferState++;
+
+                //noinspection fallthrough
+            case 2:
+                if (val.limit() != val.capacity()) {
+                    writeInt(val.limit());
+                }
+
+                if (!lastFinished) {
+                    return;
+                }
+
+                byteBufferState++;
+
+                //noinspection fallthrough
+            case 3:
+                assert !val.isReadOnly();
+
+                int length = val.capacity();
+
+                if (val.isDirect()) {
+                    long address = GridUnsafe.bufferAddress(val);
+
+                    lastFinished = writeArray(null, address, length, length);
+                } else {
+                    byte[] array = val.array();
```

Comment:
hm, didn't we agree that we should transfer only (limit - position) bytes?

Reply:
Would be weird, imo. What if receiver wants to set position to 0? These two numbers are just extra values